state/runtime/evm/state.go

```diff
@@ -210,30 +213,84 @@ func (c *state) resetReturnData() {
 
 // Run executes the virtual machine
 func (c *state) Run() ([]byte, error) {
-	var vmerr error
+	var (
+		vmerr  error
+		logged bool
+
+		op      OpCode
+		ipCopy  int
+		gasCopy uint64
+
+		ok bool
+	)
+
+	tracer := c.host.GetTracer()
+
+	defer func() {
+		if vmerr == nil || tracer == nil {
+			return
+		}
+
+		if !logged {
+			tracer.ExecuteState(
```

Comment:
Could we do something like this to avoid the many function calls?
```
type State interface {
   GetReturnData() []byte
}

type state struct {
  returnData []byte
}

func (s *state) GetReturnData() []byte {
   return s.returnData
}
```

```
type Tracer interface {
   ExecuteState(s State)
}
```

Reply:
I'm not sure whether this works because there are other several values to be passed to tracer. I added helper methods. Please check it. e803f497157cbcbfa133f82e10fd41d20ff62201